### DISCUSSION TITLE
How could I remove/Hide is_active, is_superuser and is_verified from register route?

### DISCUSSION THREAD

**Discussion by DinaTaklit** [fastapi-users/fastapi-users#882] [Q&A]
Hello the `auth/register` endpoint offer all those fields  to register the new user I want to remove/hide ` is_active`, 
`is_superuser` and `is_verified`
```python
{
  "email": "user@example.com",
  "password": "string",
  "is_active": true,
  "is_superuser": false,
  "is_verified": false,
  "firstName": "string",
  "lastName": "string",
  "phoneNumber": "string"
}
```
How is it possible to do this?

**Comment by frankie567** [comment]
Those fields are core to FastAPI Users and can't be removed.

That said, even if they appear in the Swagger documentation, they are actually *ignored* at registration. They're arbitrarily set by the backend. So you can just omit them in your API calls.

See: https://fastapi-users.github.io/fastapi-users/usage/routes/#post-register

**Reply by pplanel** [reply]
What if someone send those values in the register step? How can I ignore then?

**Reply by frankie567** [reply]
> That said, even if they appear in the Swagger documentation, they are actually **ignored at registration**. They're arbitrarily set by the backend. So you can just omit them in your API calls.

**Comment by jasonvriends** [comment]
I would not want someone to register a new account and grant themselves is_superuser. Is that set to false even if the user specifies this during creation?

**Reply by frankie567** [reply]
> That said, even if they appear in the Swagger documentation, they are actually **ignored at registration**. They're arbitrarily set by the backend. So you can just omit them in your API calls.

**Reply by felipebustillo** [reply]
> > That said, even if they appear in the Swagger documentation, they are actually **ignored at registration**. They're arbitrarily set by the backend. So you can just omit them in your API calls.

Hi Frankie, how could I achieve the same effect with extended properties on the model? For example, a new attribute "role" that always defaults to "user" and is not editable by the user. Thanks in advance!

**Reply by gonzaloverussa** [reply]
> > > That said, even if they appear in the Swagger documentation, they are actually **ignored at registration**. They're arbitrarily set by the backend. So you can just omit them in your API calls.
> 
> Hi Frankie, how could I achieve the same effect with extended properties on the model? For example, a new attribute "role" that always defaults to "user" and is not editable by the user. Thanks in advance!

You can do it by extending your "UserManager" create function like this:

```python
class UserManager(BaseUserManager[models.UC, models.UD]):
    async def create(self, user: models.UC, *args, **kwargs) -> models.UD:
        user.role = "user"
        return await super().create(user, *args, **kwargs)
```

**Comment by gonzaloverussa** [comment]
> Those fields are core to FastAPI Users and can't be removed.
> 
> That said, even if they appear in the Swagger documentation, they are actually _ignored_ at registration. They're arbitrarily set by the backend. So you can just omit them in your API calls.
> 
> See: https://fastapi-users.github.io/fastapi-users/usage/routes/#post-register

It would be nice to be able to add attributes right before the creation so they aren't in the Swagger documentation as a possible value to customize on the register endpoint.
Same with the update model. I have to add attributes to the update model so I can programmatically update them, but now they appears on the patch endpoint so I have to add them to the "create_update_dict" method.

**Reply by frankie567** [reply]
If those fields remain private, why not adding them only to the `User` model and not in the update/create schema?

**Reply by gonzaloverussa** [reply]
Hi Frankie.

Because I can't edit attributes that aren't in the update model unless I override the UserManager's update method to accept other models.
The same with the create model. I want the client to request me to create a user and be able to add or edit some attributes before the insert. I know I can do it in the on_after_register method, but this implies that the endpoint created and returned the user without the values I wanted and I have to add those attributes to the update model to update it.

I'm sorry if there is a way and I'm not seeing it.

**Reply by frankie567** [reply]
If you want to add fields programmatically right before the insertion in DB, the solution is to override [the `create` method on the `UserManager`](https://github.com/fastapi-users/fastapi-users/blob/a665cd5ed731679b342ec9c2869175447358ddba/fastapi_users/manager.py#L110-L147) and have a private `UserCreate` schema.
 
The good thing is this method returns the `User` model from the ORM (either SQLAlchemy or Beanie object). Then, you could very well add those programmatic fields to your `UserRead` schema and they will be returned in the API: you can't write them in a request, but you can read them.

```py
# schemas.py

import uuid

from fastapi_users import schemas


class UserRead(schemas.BaseUser[uuid.UUID]):
    my_private_field: str


class UserCreate(schemas.BaseUserCreate):
    pass


class UserUpdate(schemas.BaseUserUpdate):
    pass


class UserCreatePrivate(UserCreate):
    my_private_field: str
```

```py
# db.py

class User(SQLAlchemyBaseUserTableUUID, Base):
    my_private_field: str = Column(String(length=255), nullable=False)
```

```py
# manager.py

class UserManager(UUIDIDMixin, BaseUserManager[User, uuid.UUID]):
    # ...
   async def create(
        self,
        user_create: UserCreate
        safe: bool = False,
        request: Optional[Request] = None,
    ) -> User:
        user_create_private = UserCreatePrivate(**user_create.dict(), my_private_field="foo")
        return super().create(user_create_private, safe=safe, request=request)
```

**Comment by ScrimForever** [comment]
@frankie567 

I think someone can explore or "attack" endpoint, discovering is_superuser parameters, even if in the backend we set this parameter to always False, an attacker can discover this parameter through swagger. Is there any way to hide this parameter, only in the swagger?

**Comment by Housain-maina** [comment]
The is_superuser and other sensitive fields appear in the swagger docs but are actually ignored when saving a user to the database, you can modify your UserCreate model example to remove them from the swagger docs

**Comment by seanhuang514** [comment]
here is my solution
<img width="1217" alt="截圖 2024-11-26 晚上8 45 49" src="https://github.com/user-attachments/assets/7c4397bf-969c-47c0-b01b-5886cb5d8734">

```python
from fastapi_users import schemas
from pydantic import Field
from pydantic.json_schema import SkipJsonSchema

class UserRead(schemas.BaseUser[uuid.UUID]):
    # Hide fields from the response
    is_active: Optional[bool] = Field(None, exclude=True)
    is_superuser: Optional[bool] = Field(None, exclude=True)
    is_verified: Optional[bool] = Field(None, exclude=True)

class UserCreate(schemas.BaseUserCreate):
    # Hide fields from user create schema
    is_active: SkipJsonSchema[bool] = None
    is_superuser: SkipJsonSchema[bool] = None
    is_verified: SkipJsonSchema[bool] = None

router.include_router(
    fastapi_users.get_register_router(UserRead, UserCreate),
)
```
